cocotb/share/lib/simulator/simulatormodule.c

```diff
@@ -46,6 +46,161 @@ struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types */
+
+/* for gpi_sim_hdl */
+static PyTypeObject gpi_sim_hdl_Type;
+
+typedef struct {
+    PyObject_HEAD
+    gpi_sim_hdl hdl;
+} gpi_sim_hdl_Object;
+
+static PyObject *gpi_sim_hdl_repr(gpi_sim_hdl_Object *self) {
+    PyTypeObject *type = Py_TYPE(self);
+    return PyUnicode_FromFormat("<%s at %p>",
+                                  type->tp_name, self->hdl);
+}
+static Py_hash_t gpi_sim_hdl_hash(gpi_sim_hdl_Object *self) {
+    Py_hash_t ret = (Py_hash_t)self->hdl;
+    if (ret == (Py_hash_t)-1) {
+        ret = (Py_hash_t)-2;
+    }
+    return ret;
+}
+
+static PyObject *gpi_sim_hdl_richcompare(PyObject *self, PyObject *other, int op) {
+    gpi_sim_hdl_Object *self_hdl_obj;
+    gpi_sim_hdl_Object *other_hdl_obj;
+
+    if (
+            Py_TYPE(self) != &gpi_sim_hdl_Type ||
+            Py_TYPE(other) != &gpi_sim_hdl_Type) {
```

Comment:
The [documentation](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_richcompare) states 

>The first parameter is guaranteed to be an instance of the type that is defined by PyTypeObject.

So you can remove the first condition in this if, and even change the type on `self` in the parameter list.

Reply:
As written here this guards against subclasses.